{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sys.stderr = open(snakemake.log[0], \"w\")\n",
    "import pysam\n",
    "\n",
    "# see https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.cigartuples\n",
    "BAM_CEQUAL = 7\n",
    "BAM_CDIFF = 8\n",
    "BAM_CSOFT_CLIP = 4\n",
    "BAM_CINS = 1\n",
    "BAM_CDEL = 2\n",
    "\n",
    "N_WINDOW = 100\n",
    "\n",
    "\n",
    "def is_uncertain_region(record, rpos, rend, ref_fasta):\n",
    "    refseq = ref_fasta.fetch(record.reference_name, rpos - N_WINDOW, rend + N_WINDOW)\n",
    "    return (\"N\" * 10) in refseq\n",
    "\n",
    "\n",
    "def get_edit_dist(record, ref_fasta):\n",
    "    edit_dist = 0\n",
    "    qpos = 0\n",
    "    rpos = record.reference_start\n",
    "    for op, length in record.cigartuples:\n",
    "        if op == BAM_CEQUAL:\n",
    "            # no edit\n",
    "            qpos += length\n",
    "            rpos += length\n",
    "        else:\n",
    "            if op == BAM_CDIFF:\n",
    "                refseq = ref_fasta.fetch(record.reference_name, rpos, rpos + length)\n",
    "                if not is_uncertain_region(record, rpos, rpos + length, ref_fasta):\n",
    "                    # Only consider edits where the reference has a true nucleotide\n",
    "                    # because IUPAC codes lead to mason simulating an N in the reads.\n",
    "                    edit_dist += sum(base in \"ACGT\" for base in refseq)\n",
    "                qpos += length\n",
    "                rpos += length\n",
    "                    \n",
    "            elif op == BAM_CSOFT_CLIP:\n",
    "                edit_dist += length\n",
    "                qpos += length\n",
    "            elif op == BAM_CINS:\n",
    "                refseq = ref_fasta.fetch(record.reference_name, rpos, rpos + length)\n",
    "                if not is_uncertain_region(record, rpos, rpos + length, ref_fasta):\n",
    "                    # only count edit distance if the region behind the insert does not \n",
    "                    # contain N stretches in the reference. Rationale: those regions are apparently\n",
    "                    # hard to assemble, and we cannot properly simulate reads for them, so\n",
    "                    # we should not count them in a benchmark.\n",
    "                    edit_dist += length\n",
    "                qpos += length\n",
    "            elif op == BAM_CDEL:\n",
    "                refseq = ref_fasta.fetch(record.reference_name, rpos, rpos + length)\n",
    "                if not is_uncertain_region(record, rpos, rpos + length, ref_fasta):\n",
    "                    # only count edit distance if the region behind the insert does not \n",
    "                    # contain N stretches in the reference. Rationale: those regions are apparently\n",
    "                    # hard to assemble, and we cannot properly simulate reads for them, so\n",
    "                    # we should not count them in a benchmark.\n",
    "                    edit_dist += length\n",
    "                rpos += length\n",
    "            else:\n",
    "                raise ValueError(\"Unsupported CIGAR operation.\")\n",
    "    return edit_dist\n",
    "\n",
    "\n",
    "with open(snakemake.output[0], \"w\") as out:\n",
    "   print(\"Accession\", \"Contigs\", 'Total contigs', 'Contig length', 'Reference length', 'Contig frac', 'Cum frac','Relevant edit dist', 'Cigar string', 'Edit frac', sep=\"\\t\", file=out)\n",
    "   sum_of_edit_dist = 0\n",
    "   largest_no_contig = 0\n",
    "   small_larg_cover = 1.0\n",
    "   small_larg_cover_name = 'asd'\n",
    "   smallest_cum_frac = 1.0\n",
    "\n",
    "   #for bam_file in snakemake_input:\n",
    "   for bam_file, ref_fasta in zip(snakemake.input.bams, snakemake.input.refs):\n",
    "      current_contig = 1\n",
    "    \n",
    "      ref_fasta = pysam.FastaFile(ref_fasta)\n",
    "\n",
    "      with pysam.AlignmentFile(bam_file, \"rb\") as samfile:\n",
    "         total_contigs = samfile.count()\n",
    "         accession = samfile.get_reference_name(0)\n",
    "         \n",
    "         largest_no_contig = total_contigs if total_contigs > largest_no_contig else largest_no_contig\n",
    "\n",
    "      with pysam.AlignmentFile(bam_file, \"rb\") as samfile:\n",
    "         ref_lengths = samfile.lengths[0]\n",
    "         largest_coverage = 0\n",
    "         cum_frac = 0\n",
    "         for read in samfile.fetch():\n",
    "            query_alignment_length = read.query_alignment_length\n",
    "            frac = round(query_alignment_length / ref_lengths, 2)\n",
    "            \n",
    "            # Calculate edit distance from CIGAR string, because NM tag counts matching Ns as edits.\n",
    "            edit = get_edit_dist(read, ref_fasta)\n",
    "            \n",
    "            cum_frac += frac\n",
    "            cum_frac = round(cum_frac, 2)\n",
    "            sum_of_edit_dist = sum_of_edit_dist + edit\n",
    "            edit_frac = round(edit/query_alignment_length, 5)\n",
    "            largest_coverage = frac if frac > largest_coverage else largest_coverage\n",
    "\n",
    "            print(accession, current_contig, total_contigs, query_alignment_length, ref_lengths, frac, cum_frac, edit, read.cigarstring, edit_frac, sep=\"\\t\", file=out)\n",
    "\n",
    "            current_contig += 1\n",
    "         smallest_cum_frac = cum_frac if smallest_cum_frac>cum_frac else smallest_cum_frac\n",
    "         small_larg_cover = largest_coverage if small_larg_cover>largest_coverage else small_larg_cover\n",
    "         small_larg_cover_name = accession if small_larg_cover>=largest_coverage else small_larg_cover_name\n",
    "\n",
    "   print('Largest number of contigs')\n",
    "   print(largest_no_contig)\n",
    "   print('Smallest largest coverage in ' + str(small_larg_cover_name))\n",
    "   print(small_larg_cover)\n",
    "   print('Sum of edit distances')\n",
    "   print(sum_of_edit_dist)\n",
    "   print('Smallest cum. fraction of contigs')\n",
    "   print(smallest_cum_frac)\n",
    "   print('Largest number of contigs', file=out)\n",
    "   print(largest_no_contig, file=out)\n",
    "   print('Smallest largest coverage in ' + str(small_larg_cover_name), file=out)\n",
    "   print(small_larg_cover, file=out)\n",
    "   print('Sum of edit distances', file=out)\n",
    "   print(sum_of_edit_dist, file=out)\n",
    "   print('Smallest cum. fraction of contigs', file=out)\n",
    "   print(smallest_cum_frac, file=out)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
